source/sigv4.c

```diff
@@ -793,6 +854,230 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
     }
 
 /*-----------------------------------------------------------*/
+    static SigV4Status_t writeSignedHeaderToString( size_t headerIndex,
+                                                    uint32_t flags,
+                                                    CanonicalContext_t * canonicalRequest )
+    {
+        char * pBufLoc;
+        size_t buffRemaining, keyLen = 0, i = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        pBufLoc = canonicalRequest->pBufCur;
+        buffRemaining = canonicalRequest->bufRemaining;
+
+        keyLen = canonicalRequest->pHeadersLoc[ headerNum ].key.dataLen;
+
+        if( keyLen + 1 > buffRemaining )
+        {
+            sigV4Status = SigV4InsufficientMemory;
+        }
+        else
+        {
+            for( i = 0; i < keyLen; i++ )
+            {
+                if( !( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) )
+                {
+                    *pBufLoc = tolower( canonicalRequest->pHeadersLoc[ headerNum ].key.pData[ i ] );
```

Comment:
The header field/key must also be trimmed of whitespace

Reply:
The following pseudocode describes how to construct the canonical list of headers:
```
CanonicalHeaders =
CanonicalHeadersEntry0 + CanonicalHeadersEntry1 + ... + CanonicalHeadersEntryN
CanonicalHeadersEntry =
Lowercase(HeaderName) + ':' + Trimall(HeaderValue) + '\n'
```
As per the AWS doc it does not mention about trimming header key.

Reply:
To create the canonical headers list, **convert all header names to lowercase and remove leading spaces and trailing spaces**. Convert sequential spaces in the header value to a single space.

Reply:
Is this function writing "SignedHeaders" or "CanonicalHeaders" to the CanonicalRequest? The following is the formula of the CanonicalRequest where these 2 are distinct elements of the equation:
```
CanonicalRequest =
  HTTPRequestMethod + '\n' +
  CanonicalURI + '\n' +
  CanonicalQueryString + '\n' +
  CanonicalHeaders + '\n' +
  SignedHeaders + '\n' +
  HexEncode(Hash(RequestPayload))
  ```

Reply:
This is adding Signed Headers to the canonical Request. Although the name mentions about adding Signed headers to the cancel request. I will rename it again so as to make it more comprehnsible.